### ISSUE TITLE
Memory leak in this code, using Image.merge()

### ISSUE THREAD

**Issue by VOTProductions** [python-pillow/Pillow#1187]
Hello,

I recently have encountered a memory leak in this code:
See code in next post.

I got this code from:
https://stackoverflow.com/questions/29332424/python-3-changing-colour-of-an-image/

It seems that the RAM usage increases by about 50-150kb every time I run my code. The reason for all these for loops is to speed up the leak as my real program uses this tint code many times throughout its life. The person who made this code has suggested that this is possibly a leak in Pillow. I feel this is a leak as the memory usage keeps rising and generally doesn't stop. When the input line is called, the memory usage stays still.

Python 3.4.3, Pillow 2.8.1.

**Comment by VOTProductions** [comment]
I updated my code to be less cluttered:
```
from PIL import Image
from PIL.ImageColor import getcolor, getrgb
from PIL.ImageOps import grayscale
import time
import gc
for i in range(1,40):
    time.sleep(0.2)
    def image_tint(src, tint='#ffffff'):
        if Image.isStringType(src):  # file path?
            src = Image.open(src)
        if src.mode not in ('RGB', 'RGBA'):
            raise TypeError('Unsupported source image mode: {}'.format(src.mode))
        src.load()
         tr, tg, tb = getrgb(tint)
        tl = getcolor(tint, "L")  # tint color's overall luminosity
        if not tl: tl = 1  # avoid division by zero
        tl = float(tl)  # compute luminosity preserving tint factors
        sr, sg, sb = map(lambda tv: tv/tl, (tr, tg, tb))  # per component
                                                          # adjustments
        # create look-up tables to map luminosity to adjusted tint
        # (using floating-point math only to compute table)
        luts = (tuple(map(lambda lr: int(lr*sr + 0.5), range(256))) +
                tuple(map(lambda lg: int(lg*sg + 0.5), range(256))) +
                tuple(map(lambda lb: int(lb*sb + 0.5), range(256))))
        l = grayscale(src)  # 8-bit luminosity version of whole image
        if Image.getmodebands(src.mode) < 4:
            merge_args = (src.mode, (l, l, l))  # for RGB verion of grayscale
        else:  # include copy of src image's alpha layer
            a = Image.new("L", src.size)
            a.putdata(src.getdata(3))
            merge_args = (src.mode, (l, l, l, a))  # for RGBA verion of grayscale
            luts += tuple(range(256))  # for 1:1 mapping of copied alpha values
        potato = Image.merge(*merge_args).point(luts)
        src.close()
        return potato
    if __name__ == '__main__':
        for i in range(1,5):
            import os
            input_image_path = 'headbase.png' # any .png will do
            root, ext = os.path.splitext(input_image_path)
            result_image_path = root+'_result_py3'+ext
            result = image_tint(input_image_path, '#383D2D')
            if os.path.exists(result_image_path):  # delete any previous result file
                os.remove(result_image_path)
            result.save(result_image_path)  # file name's extension determines format
            gc.collect()
input("test")
```

Any .png should exhibit this issue, but this is headbase.png that I'm using: http://i.imgur.com/z3pKTOY.png

**Comment by homm** [comment]
I confirm. The problem in `putdata`. The minimal test case is:

```python
from PIL import Image
import gc

i = Image.new('RGB', (25, 25))
data = i.getdata()

for _ in range(5000):
    i.putdata(data)
    gc.collect()

input("test")
```

**Comment by gareth-rees** [comment]
I can reproduce the leak (OS X 10.10.3, Python 3.4.3, MacPorts Pillow 2.7.0). Here's a much-reduced version of the test case that still exhibits the leak:
```
import gc
import os
from PIL import Image

def main(filename='test.png'):
    Image.new('RGBA', (100, 100)).save(filename)
    for i in range(10):
        # This works on BSD -- might need adjusting on Linux
        os.system('ps -o rss= {}'.format(os.getpid()))
        img = Image.open(filename)
        img.load()
        a = Image.new("L", img.size)
        a.putdata(img.getdata(3))
        img.close()
        gc.collect()

if __name__ == '__main__':
    main()        
```
The culprit would seem to be the call to `putdata`.

**Comment by gareth-rees** [comment]
The [implementation of `putdata`](https://github.com/python-pillow/Pillow/blob/master/_imaging.c#L1217) calls:
```
           seq = PySequence_Fast(data, must_be_sequence);
```
in two places. `PySequence_Fast` increments the reference count of its result, so there needs to be a corresponding call to `Py_DECREF(seq)`. But this is missing (in both cases), so `seq` is leaked.

**Comment by hugovk** [comment]
Fixed in #1196.
